src/vm/methodtablebuilder.cpp

```diff
@@ -9519,42 +9519,115 @@ void MethodTableBuilder::CheckForSystemTypes()
 {
     STANDARD_VM_CONTRACT;
 
+    LPCUTF8 name, nameSpace;
+
     MethodTable * pMT = GetHalfBakedMethodTable();
     EEClass * pClass = GetHalfBakedClass();
 
     // We can exit early for generic types - there are just a few cases to check for.
-    if (bmtGenerics->HasInstantiation() && g_pNullableClass != NULL)
+    if (bmtGenerics->HasInstantiation() && pClass->HasLayout())
     {
-        _ASSERTE(g_pByReferenceClass != NULL);
-        _ASSERTE(g_pByReferenceClass->IsByRefLike());
+        if (FAILED(GetMDImport()->GetNameOfTypeDef(GetCl(), &name, &nameSpace)))
```

Comment:
> Are there plans to support vector types defined in OOB assemblies

No, only supporting the types in corelib.

> Why do we do a name match here instead of keeping references to these in globals

Because we were already name matching these everywhere else. I can update to keep references to them instead.

Reply:
Do we want all the same handling as `g_pNullableClass`/`g_pByReferenceClass`? Nullable and ByRef  are specially bound in `SystemDomain::LoadBaseSystemClasses`.

Reply:
It is always nice to do the name checks only when they are really required. Can we do the name check here for intrinsic types only?

Reply:
@jkotas, are you indicating we should keep the current logic, rather than making it do what `g_pNullableClass` is doing?

Reply:
I was suggesting that the name check on the generic path can be done for intrinsic types only. It is what we generally do elsewhere - check the names for intrinsic types only.

Reply:
Sorry. I guess I'm still not completely clear on whether you want changes (and if you do, what they are), or if the way I have it now is sufficient (and we should revisit if/when core types, like string, take a dependency on hardware intrinsics).

Reply:
- Move the check for the intrinsic attribute earlier so that we now whether the type is intrinsic here.
- Check the name of the generic type here only when the type is intrinsic.

Reply:
@jkotas, it looks like this would require moving the `SetIsIntrinsicType` code into `SetupMethodTable2`.

Right now it is in `BuildMethodTableThrowing` a few lines after ([SetupMethodTable2](https://github.com/dotnet/coreclr/blob/master/src/vm/methodtablebuilder.cpp#L1834) is L1834, we get the [HalfBakedMethodTable](https://github.com/dotnet/coreclr/blob/master/src/vm/methodtablebuilder.cpp#L1839) on L1839 and then do the [intrinsic checks](https://github.com/dotnet/coreclr/blob/master/src/vm/methodtablebuilder.cpp#L1866) on L1866).

So we would basically end up moving the check to just before the call to `CheckForSystemTypes()` on https://github.com/dotnet/coreclr/blob/master/src/vm/methodtablebuilder.cpp#L10449, does that sound correct?

Reply:
Sounds fine to me.

Reply:
Should be handled now.